### DISCUSSION TITLE
Vercel Blob Storage Plugin client upload only files under 4,5 mb possible to upload

### DISCUSSION THREAD

**Discussion by christianfullstack** [payloadcms/payload#7569] [Roadmap]
### EDIT: Available in v3.25.0

🎉 🎉 🎉 🎉  Now available in https://github.com/payloadcms/payload/releases/tag/v3.25.0

### Payload Version

v3.0.0-beta.73

### Node Version

20.9.0

### Next.js Version

15.0.0-canary.83

### Describe the Bug

at the moment the @payloadcms/storage-vercel-blob allows only uploads smaller than 4,5mb.

Could we switch to client upload to allow bigger filesizes ? I think we need the client upload.
[https://vercel.com/docs/storage/vercel-blob/client-upload#client-uploads-with-vercel-blob](https://vercel.com/docs/storage/vercel-blob/client-upload#client-uploads-with-vercel-blob)

because this limit is not useful in the case of using vercel blob storage.
`upload: {
    limits: {
      fileSize: 30000000,
    },
  },`

Thank you

### Reproduction Steps

use @payloadcms/storage-vercel-blob and try to upload a file bigger than 5 mb.

### Adapters and Plugins

storage-vercel-blob

**Comment by jmikrut** [comment]
Hey @christianfullstack — are you deploying on Vercel? I know the 4.5mb thing is a Vercel limitation, but I'd like to confirm.

We do indeed have plans to switch to client-side uploads so I have added this to our Roadmap!

**Reply by christianfullstack** [reply]
Hey @jmikrut,
yes i deploy to vercel and i have setup the blob storage in the database tab of my site.

i was wondering why the config Object has the prop access: 'public'.

`vercelBlobStorage({
      access: 'public',
      collections: {
        [Media.slug]: true,
      },
      token: process.env.BLOB_READ_WRITE_TOKEN || '',
    }),`

Thank you

**Comment by robinscholz** [comment]
@jmikrut roughly how far out is this feature? We are close to deploying a larger project and this issue would prevent us from hosting it on Vercel sadly.

**Comment by nico-k3** [comment]
Yea, we have had some problems because of that too. Deployed 3 projects with 3.0 on Vercel so far with 2 more in active development and this is the biggest complaint from the clients. It is hard to communicate to them, that they have to downsize their images before uploading them to their CMS.

We're using S3 tho, so would be really nice if this feature would be available there too.

**Comment by drewlonious** [comment]
is there an update on this? using s3/supabase, mono repo deployed to vercel. seems to defeat the purpose of the monorepo

**Comment by hansfuchs** [comment]
I'm also using Vercel + Supabase, wish I had known about this limitation before building the whole CMS for a customer. Is there still no update in regards to this, @jmikrut ?

**Comment by ydemetriades** [comment]
Do we have any updates on this?

**Comment by noahwaldner** [comment]
I was just looking at this today and was able to get a direct upload to S3 working. The downside is that it skips all of payloads image processing such as cropping or splitting up the image into different resolutions.

One option would be to process the images after the initial upload with a kind of worker. This would need to happen off Vercel, as else the 4.5MB Limit would also kick in there.

@jmikrut Did you decide on how to handle this? 

I'm happy to contribute an implementation for this!

**Reply by mwdossantos** [reply]
Hey @noahwaldner, how did you get the direct upload working?

**Reply by noahwaldner** [reply]
Hey @mwdossantos i create signed urls to upload from inside payload and use it on the client to directly upload the file to s3. 

This solution bypasses `sharp` inside payload, so you'll only have the original image size without optimization

**Comment by giorgiPapava** [comment]
Any updates on this? I have same problem on s3 bucket deployed with vercel

**Comment by altarapid** [comment]
Are there any updates?

**Comment by jeromewtf** [comment]
Would love to hear updates :)

**Comment by gabrielkuettel** [comment]
ditto :)

**Comment by freeman-jiang** [comment]
sadly wandered upon this issue like everyone else :(

would really appreciate client-side upload—i think it could make sense if it uploads the main object directly to the bucket first.

**Reply by noahwaldner** [reply]
Would a plugin for direct upload that lacks the ability to resize images be of any benefit? You'd basically always store the full res image.

**Reply by klapec** [reply]
@noahwaldner absolutely! we're currently facing this issue when attempting to upload a bigger (~20MB) file using the s3 adapter. would you mind sharing your plugin?

**Comment by christophemenager** [comment]
Same issue here, I am trying to upload mp3 and mp4 files which are very often above the 4.5 Mb limitation. I am quite surprised to discover this once it's in production: it's not mentioned anywhere in the doc and it seems like a real pity :/

**Reply by christophemenager** [reply]
For info, I ended up deploying on Digital Ocean App Platform for 10$/month. It's also much faster as it avoids Edge functions cold starts. But I wish I would have read this somewhere in the docs

**Comment by miguderp** [comment]
Would "resumable uploads" be a viable path? I'm not too savy on the subject but [Supabase did this](https://supabase.com/docs/guides/storage/uploads/resumable-uploads)

**Comment by Bananenbrot1** [comment]
Would love to have this feature. Any news on the state here?

**Comment by Hworden** [comment]
Just to add a note: I'm running into a similar issue but on AWS using open-next.
There I'm hitting the [6MB invocation payload limit](https://docs.aws.amazon.com/lambda/latest/dg/gettingstarted-limits.html)

**Comment by XGZepto** [comment]
Is there any update on client uploads? Wish I have come across this discussion earlier :(

**Comment by lynndylanhurley** [comment]
Want to also voice how frustrating this issue is. We've launched several projects with payload over the last year, and this has been a problem on literally every single one of them. It's embarrassing to explain to clients in 2025 that we can only support 4.5MB video uploads.

**Comment by GodGardensGuns** [comment]
My solution was to host Payload on a VPS (like DigitalOcean, Hetzner, Linode, etc.) using Coolify. You could call it "self-hosted" but Coolify takes care of most of the headaches in that department. My client can now upload files of any size and the server only costs around $10/mo.

Hosting on Vercel is extremely convenient but limitations like this are a no-go.

I will add that this issue should be added to the docs until it's resolved. It's not the best feeling to have to switch hosting platforms right before launch while on a deadline.

**Reply by iamlinkus** [reply]
I completely agree. I think not having this limitation within the docs (specifically in the part about uploads) is counter productive, especially since a huge chunk of new payload developers will be hosting on vercel primarily. 

Last year we launched 3 huge websites almost at the same time with payload 3.0 and we too encountered this limitation completely on accident when there were 2 days left before we had to launch. Now the client has been forced to work around this by uploading bigger files straight to S3 and then pasting the file url in payload, which is very inconvenient, but that was the only way we could launch in time...

**Reply by christophemenager** [reply]
@wjd3 agreed! I solved this issue by self-hosting on hetzner + coolify.

The main pain point was to discover this last minute.

PayloadCMS marketing is based a lot on their nextjs move with the v3, but they forget to mention this "little detail" that you can't really use it in "real world" where we have to store large image, sounds or video like any other CMS on the market would offer :)

**Reply by r1tsuu** [reply]
> PayloadCMS marketing is based a lot on their nextjs move with the v3, but they forget to mention this "little detail" that you can't really use it in "real world" where we have to store large image, sounds or video like any other CMS on the market would offer :)

We'll be able to do large uploads on Vercel very soon, you can track the PR here https://github.com/payloadcms/payload/pull/11382!

**Comment by r1tsuu** [comment]
The PR is here - https://github.com/payloadcms/payload/pull/11382.

**Reply by denolfe** [reply]
Now available in https://github.com/payloadcms/payload/releases/tag/v3.25.0

**Comment by robertmalicke** [comment]
Hi Amazing work. Is it possible we could get a tiny bit of guidance on how to allow CORS on Vercel Blob? I have added the following policy in my next config to no avail

```
  async headers() {
    return [
      {
        source: '/:path*',
        headers: [
          { key: 'Access-Control-Allow-Credentials', value: 'true' },
          { key: 'Access-Control-Allow-Origin', value: '*' },
          {
            key: 'Access-Control-Allow-Methods',
            value: 'GET,OPTIONS,PATCH,DELETE,POST,PUT',
          },
          {
            key: 'Access-Control-Allow-Headers',
            value:
              'X-CSRF-Token, X-Requested-With, Accept, Accept-Version, Content-Length, Content-MD5, Content-Type, Date, X-Api-Version',
          },
        ],
      },
    ]
  },
  ```
  
  This could be a good start of documentation on the Payload docs website.
